srml/contracts/src/wasm/runtime.rs

```diff
@@ -186,12 +208,32 @@ fn read_sandbox_memory<E: Ext>(
 ) -> Result<Vec<u8>, sandbox::HostError> {
 	charge_gas(ctx.gas_meter, ctx.schedule, RuntimeToken::ReadMemory(len))?;
 
-	let mut buf = Vec::new();
-	buf.resize(len as usize, 0);
+	let mut buf = vec![0u8; len as usize];
+	ctx.memory().get(ptr, buf.as_mut_slice()).map_err(|_| sandbox::HostError)?;
+	Ok(buf)
+}
 
-	ctx.memory().get(ptr, &mut buf)?;
+/// Read designated chunk from the sandbox memory into the scratch buffer, consuming an
+/// appropriate amount of gas. Resizes the scratch buffer to the specified length on success.
+///
+/// Returns `Err` if one of the following conditions occurs:
+///
+/// - calculating the gas cost resulted in overflow.
+/// - out of gas
+/// - requested buffer is not within the bounds of the sandbox memory.
+fn read_sandbox_memory_into_scratch<E: Ext>(
+	ctx: &mut Runtime<E>,
+	ptr: u32,
+	len: u32,
+) -> Result<(), sandbox::HostError> {
+	charge_gas(ctx.gas_meter, ctx.schedule, RuntimeToken::ReadMemory(len))?;
 
-	Ok(buf)
+	let mut buf = mem::replace(&mut ctx.scratch_buf, Vec::new());
+	buf.resize(len as usize, 0);
+	ctx.memory().get(ptr, buf.as_mut_slice()).map_err(|_| sandbox::HostError)?;
+	let _  = mem::replace(&mut ctx.scratch_buf, buf);
```

Comment:
The reason you use replace is to satisfy the property: `Resizes the scratch buffer to the specified length on success.` ?
Is this property very usefull ? because it seems to me we could avoid this Vec::new() here by resizing any way. do I miss something ?

Reply:
The reason for the first replace is because of the borrow checker. We need a mutable reference to the scratch buffer and need to borrow `ctx` for `memory()`.

The second `replace` could be switched to `ctx.scratch_buf = buf;`, which is a bit more obvious.

Reply:
can't you just access memory by the field directly ?

Reply:
Yeah, good point. Will change.